### PR TITLE
[Model] Implemented PGExplainer for Heterogeneous graph

### PR DESCRIPTION
**Pull request by ndbaker1**
## Description

Implemented PGExplainer (https://arxiv.org/pdf/2011.04573.pdf) for Heterogeneous Graphs.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes

Added a new class HeteroPGExplainer for Heterogeneous graph explanation.

The output of the model has been manually verified on MUTAG dataset. A visualization script was created to visualize the mutag dataset (external ref: https://gist.github.com/ndbaker1/68b956f9232acc948116e90d3aad14af). Unfortunately, there's no ground truth for the explanation, so we cannot quantify the model's performance. Instead, we manually investigated the results and confirmed that they aligned well with human intuition.

Inline documentation is added following the original style, and an example of using the module is included.

### CODE REVIEW

**Comment by mufeili** [comment]
I have a high-level question. Is it sufficient to call `to_homogeneous` first and then use the previous PGExplainer implementation for homogeneous graphs? If not, what will be the gaps?

**Comment by ndbaker1** [comment]
> I have a high-level question. Is it sufficient to call `to_homogeneous` first and then use the previous PGExplainer implementation for homogeneous graphs? If not, what will be the gaps?

I believe the gap in this approach is that the Heterograph classifier will not be usable by existing PGExplainer when performing model evaluations

The approach I've gone with has been to store the edge mask in its homograph format in order to keep the internals of PGExplainer intact, and use the context from the `EIDS` data to reobtain the heterograph edge mask

**Comment by mufeili** [comment]
@dgl-bot

File: python/dgl/nn/pytorch/explain/pgexplainer.py [hunk 1]
```diff
@@ -358,3 +360,382 @@ def explain_graph(self, graph, feat, tmp=1.0, training=False, **kwargs):
         self.clear_masks()
 
         return (probs, edge_mask) if training else (probs.data, edge_mask)
+
+
+class HeteroPGExplainer(nn.Module):
```
[the remark this answers is not in the document]
Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:365]: makes sense, it should save lots on repetition since only `explain_graph` and `train_step` require changes

[the remark this answers is not in the document]
Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:365]: before this, do you approve of removing the `feat` parameter completely from [`set_masks`](https://github.com/dmlc/dgl/blob/147f6d243532e52c082623463d69cf444030112c/python/dgl/nn/pytorch/explain/pgexplainer.py#L67-L85) in `PGExplainer` and replacing the `num_nodes` lines with:

```python
num_nodes, _ = feat.shape
```

with

```python
num_nodes = graph.num_nodes()
```

this would make the function usable within the hetero impl. as well

[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:365]: Sure.

[the remark this answers is not in the document]
Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py]: done 👍

**Comment by mufeili** [comment]
@dgl-bot

**Comment by mufeili** [comment]
Lint check failed

```
python/dgl/nn/pytorch/explain/pgexplainer.py:391:4: W0235: Useless super delegation in method '__init__' (useless-super-delegation)
```

File: python/dgl/nn/pytorch/explain/pgexplainer.py [hunk 2]
```diff
@@ -358,3 +357,243 @@ def explain_graph(self, graph, feat, tmp=1.0, training=False, **kwargs):
...
+        ...
+        ...         if embed:
+        ...             return h
+        ...
+        ...         with g.local_scope():
+        ...             g.ndata["h"] = h
+        ...             hg = 0
+        ...             for ntype in g.ntypes:
+        ...                 hg = hg + dgl.mean_nodes(g, "h", ntype=ntype)
+        ...             return th.sigmoid(self.fc(hg))
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/pgexplainer.py:512]: Using `th.sigmoid` here may not be appropriate. In particular, `out_feats` is not necessarily 2.

File: python/dgl/nn/pytorch/explain/pgexplainer.py [hunk 3]
```diff
@@ -358,3 +357,243 @@ def explain_graph(self, graph, feat, tmp=1.0, training=False, **kwargs):
...
+        super(HeteroPGExplainer, self).__init__(
+            model,
+            num_features,
+            coff_budget,
+            coff_connect,
+            sample_bias,
+        )
+
+    def train_step(self, graph, feat, tmp, **kwargs):
+        r"""Compute the loss of the explanation network
```
[the remark this answers is not in the document]
Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:408]: cool, done

File: python/dgl/nn/pytorch/explain/pgexplainer.py (hunk 2 above)
Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:512]: you are right I forgot to fix this, done now.

File: python/dgl/nn/pytorch/explain/pgexplainer.py [hunk 4]
```diff
@@ -358,3 +356,227 @@ def explain_graph(self, graph, feat, tmp=1.0, training=False, **kwargs):
...
+        ...                 etype: {"edge_weight": mask} for etype, mask in edge_weight.items()
+        ...             }
+        ...             h = self.conv(g, h, mod_kwargs=mod_kwargs)
+        ...         else:
+        ...             h = self.conv(g, h)
+        ...
+        ...         if embed:
+        ...             return h
+        ...
+        ...         with g.local_scope():
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/pgexplainer.py:490]: Could you also update the example for homogeneous graphs regarding the use of `if embed:` and `with g.local_scope():`?

File: python/dgl/nn/pytorch/explain/pgexplainer.py [hunk 5]
```diff
@@ -358,3 +356,227 @@ def explain_graph(self, graph, feat, tmp=1.0, training=False, **kwargs):
...
+        >>> num_classes = 2
+        >>> g = dgl.heterograph({("user", "plays", "game"): ([0, 1, 1, 2], [0, 0, 1, 1])})
+        >>> g.nodes["user"].data["h"] = th.randn(g.num_nodes("user"), input_dim)
+        >>> g.nodes["game"].data["h"] = th.randn(g.num_nodes("game"), input_dim)
+
+        >>> transform = dgl.transforms.AddReverse()
+        >>> g = transform(g)
+
+        >>> # define and train the model
+        >>> model = Model(input_dim, hidden_dim, num_classes, g.canonical_etypes)
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/pgexplainer.py:509]: Maybe you don't need `hidden_dim` for a demo. Simply using `input_dim` can be sufficient.

File: tests/python/pytorch/nn/test_nn.py [hunk 6]
```diff
@@ -1839,6 +1839,64 @@ def forward(self, g, h, embed=False, edge_weight=None):
     probs, edge_weight = explainer.explain_graph(g, feat)
 
 
+@pytest.mark.parametrize("g", get_cases(["hetero"], exclude=["zero-degree"]))
```
Comment (mufeili) [inline tests/python/pytorch/nn/test_nn.py:1842]: Does this test cover batched graphs?

**Comment by mufeili** [comment]
@dgl-bot

File: python/dgl/nn/pytorch/explain/pgexplainer.py [hunk 7]
```diff
@@ -358,3 +356,227 @@ def explain_graph(self, graph, feat, tmp=1.0, training=False, **kwargs):
...
+        self.elayers = self.elayers.to(graph.device)
+
+        pred = self.model(graph, feat, embed=False, **kwargs).argmax(-1).data
+
+        prob, hetero_edge_mask = self.explain_graph(
+            graph, feat, tmp=tmp, training=True, **kwargs
+        )
+
+        # store edge mask in homogeneous format
+        self.edge_mask = torch.concat(list(hetero_edge_mask.values()))
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/pgexplainer.py:422]: Is this guaranteed to be the order of the node/edge types between the conversion of homogeneous graphs and heterogeneous graphs?

File: python/dgl/nn/pytorch/explain/pgexplainer.py [hunk 8]
```diff
@@ -358,3 +356,227 @@ def explain_graph(self, graph, feat, tmp=1.0, training=False, **kwargs):
...
+
+        >>> # Explain the graph
+        >>> feat = g.ndata.pop("h")
+        >>> probs, edge_mask = explainer.explain_graph(g, feat)
+        """
+        self.model = self.model.to(graph.device)
+        self.elayers = self.elayers.to(graph.device)
+
+        embed = self.model(graph, feat, embed=True, **kwargs)
+        embed_flat = torch.concat(list(embed.values()))
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/pgexplainer.py:540]: ditto as above

Comment (mufeili) [inline python/dgl/nn/pytorch/explain/pgexplainer.py:540]: For safely, it might be better to detach the embeddings from the computation graph by `.data`.

File: python/dgl/nn/pytorch/explain/pgexplainer.py [hunk 9]
```diff
@@ -358,3 +356,227 @@ def explain_graph(self, graph, feat, tmp=1.0, training=False, **kwargs):
...
+        values = emb.reshape(-1)
+
+        values = self.concrete_sample(values, beta=tmp, training=training)
+        self.sparse_mask_values = values
+
+        reverse_eids = homo_graph.edge_ids(row, col).long()
+        edge_mask = (values + values[reverse_eids]) / 2
+
+        self.clear_masks()
+        self.set_masks(homo_graph, edge_mask)
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/pgexplainer.py:560]: Do you still need to call `self.clear_masks()` and `self.set_masks` inside `explain_graph`? Is it sufficient to call them inside `train_step`?

**Review by mufeili — COMMENTED** [review]
done a pass, some comments

File: python/dgl/nn/pytorch/explain/pgexplainer.py (hunk 4 above)
Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:490]: sure thing, I will standardize the usage of this across examples and tests

File: tests/python/pytorch/nn/test_nn.py (hunk 6 above)
Comment (ndbaker1) [reply tests/python/pytorch/nn/test_nn.py:1842]: sorry has me a bit confused, should I try to cover all possible heterographs in `get_cases`?

something like this:
```suggestion
@pytest.mark.parametrize("g", get_cases(["hetero", "bipartite", "block"], exclude=["block-bipartite"]))
```

File: python/dgl/nn/pytorch/explain/pgexplainer.py (hunk 8 above)
Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:540]: done

Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:540]: I have updated it to let `to_homogeneous` handle the merge like you mentioned

File: python/dgl/nn/pytorch/explain/pgexplainer.py (hunk 7 above)
Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:422]: Initially I relied on the determinism of the edge dict, but based on the convs. below I think it makes more sense to retain `edge_mask` from the `explain_graph` step and avoid the redundant conversion.

let me know what you think of the changes once they are visible

File: python/dgl/nn/pytorch/explain/pgexplainer.py (hunk 9 above)
Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:560]: I don't see the need for the `clear_masks` calls as well.

with respect to the changes related to an [above comment](https://github.com/dmlc/dgl/pull/5739#discussion_r1214815720), I can keep the final clear based on the `training` param.

Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:547]: I also think the `clear_masks` call is only necessary when there is no training, so following changes from [convo. above](https://github.com/dmlc/dgl/pull/5739#discussion_r1214815720) I think it should suffice to clear the mask when it is not in training setting

Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:547]: I don't personally see any cases where `set_masks` is necessary inside the class because the steps from `explain_graph` shouldn't yeild a `None` for `edge_mask`?, but don't want to be too eager about pruning code.

File: tests/python/pytorch/nn/test_nn.py (hunk 6 above)
Comment (mufeili) [reply tests/python/pytorch/nn/test_nn.py]: From the documentation you added, the module supports a batched graph, one that is created with [dgl.batch](https://docs.dgl.ai/generated/dgl.batch.html#dgl.batch). However, previously the unit test did not cover a batched heterogeneous graph. I was suggesting the expansion of the unit test to cover batched heterogeneous graphs. You don't necessarily need to use `get_cases`. It's fine to directly create graphs inside the unit test function.

File: python/dgl/nn/pytorch/explain/pgexplainer.py (hunk 7 above)
Comment (mufeili) [reply python/dgl/nn/pytorch/explain/pgexplainer.py]: What if we reply on `to_homogeneous` to perform this kind of conversion? In this way, it's guaranteed that the order of edge-specific information is consistent with the result of `to_homogeneous`.

File: python/dgl/nn/pytorch/explain/pgexplainer.py [hunk 10]
```diff
@@ -348,13 +346,225 @@ def explain_graph(self, graph, feat, tmp=1.0, training=False, **kwargs):
...
+            The temperature parameter fed to the sampling procedure.
+        kwargs : dict
+            Additional arguments passed to the GNN model.
+
+        Returns
+        -------
+        Tensor
+            A scalar tensor representing the loss.
+        """
+        return super().train_step(graph, feat, tmp=tmp, **kwargs)
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/pgexplainer.py:412]: Do you still need L390-412 then?

File: python/dgl/nn/pytorch/explain/pgexplainer.py (hunk 7 above)
Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py]: will make sure the conversions are always obtained using `to_homogeneous` 👍

File: tests/python/pytorch/nn/test_nn.py (hunk 6 above)
Comment (ndbaker1) [reply tests/python/pytorch/nn/test_nn.py]: oh i understand now, sorry for the confusion. It makes sense to me to keep a case for this so let me try to make an addition to `get_cases`

Comment (ndbaker1) [reply tests/python/pytorch/nn/test_nn.py]: added

File: python/dgl/nn/pytorch/explain/pgexplainer.py (hunk 10 above)
Comment (mufeili) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:412]: I see. I don't have a better solution. Let's keep it as so for now.

**Comment by mufeili** [comment]
@dgl-bot

File: python/dgl/nn/pytorch/explain/pgexplainer.py [hunk 11]
```diff
@@ -348,13 +346,225 @@ def explain_graph(self, graph, feat, tmp=1.0, training=False, **kwargs):
...
+        >>> for epoch in range(20):
+        ...     tmp = float(init_tmp * np.power(final_tmp / init_tmp, epoch / 20))
+        ...     loss = explainer.train_step(g, g.ndata["h"], tmp)
+        ...     optimizer_exp.zero_grad()
+        ...     loss.backward()
+        ...     optimizer_exp.step()
+
+        >>> # Explain the graph
+        >>> feat = g.ndata.pop("h")
+        >>> probs, edge_mask = explainer.explain_graph(g, feat)
```
Comment (mufeili) [inline python/dgl/nn/pytorch/explain/pgexplainer.py:521]: One last comment. I got 

```
{('game',
  'rev_plays',
  'user'): tensor([3.0763e-04, 1.3422e-04, 9.6319e-07, 2.5353e-05],
        grad_fn=<IndexBackward0>),
 ('user',
  'plays',
  'game'): tensor([3.0763e-04, 1.3422e-04, 9.6319e-07, 2.5353e-05],
        grad_fn=<IndexBackward0>)}
```

for `edge_mask`. In other words, the reverse edges have the same importance as the forward edges. Is this expected from the perspective of `PGExplainer`?

Comment (kunmukh) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:521]: `
To explain predictions made by a GNN model, we divide the original input graph Go into two subgraphs: Go = Gs + ∆G, where Gs presents the underlying subgraph that makes important contributions to GNN’s predictions, which is the expected explanatory graph, and ∆G consists of the remaining task-irrelevant edges for predictions made by the GNN
`

From the above line in the paper, I understand that when we select an edge as important (e.g., be part of `Gs`), we also select its reverse edge.  Therefore, the reverse edges are expected to have the same importance as the forward edges of the selected edges.

Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:521]: let me review the reference material for any other possible remarks

Comment (ndbaker1) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:521]: code-wise this is a result of the following logic within the `explain_graph` call, which obtains symmetric edge weights:

```python
reverse_eids = homo_graph.edge_ids(row, col).long()
edge_mask = (values + values[reverse_eids]) / 2
```

If I had to rationalize this, I would say it is due to the implicit requirement that every node has a corresponding embedding, which explains the addition of reverse-edges in order to guarantee global message-passing coverage. Then as a result we assume that these edges share the same score?

Comment (mufeili) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:521]: @ndbaker1 @kunmukh Is this from a prior implementation? I did not find this in the paper.

Comment (kunmukh) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:521]: In the original implementation, https://github.com/flyingdoog/PGExplainer/blob/master/codes/Explainer.py#L27, I saw this line `sym_mask = (sym_mask + tf.transpose(sym_mask)) / 2`, where they calculate a symmetric mask by adding sym_mask with its transpose and dividing the result by 2. This ensures that the mask is symmetric along the diagonal.

Since the mask is symmetric along the diagonal, the edges and reverse edges have `edge_mask = (values + values[reverse_eids]) / 2` . 

@ndbaker1, please correct me if I missed something.

Comment (mufeili) [reply python/dgl/nn/pytorch/explain/pgexplainer.py:521]: I see. Thanks. I'm good then.

**Review by mufeili — APPROVED** [review]
